File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/firestore/utils/PipelineStageHandlers.java
/*
 * Copyright 2026, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.firestore.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Pipeline;
import com.google.firebase.firestore.pipeline.AggregateOptions;
import com.google.firebase.firestore.pipeline.AggregateStage;
import com.google.firebase.firestore.pipeline.AliasedAggregate;
import com.google.firebase.firestore.pipeline.BooleanExpression;
import com.google.firebase.firestore.pipeline.Expression;
import com.google.firebase.firestore.pipeline.Field;
import com.google.firebase.firestore.pipeline.FindNearestOptions;
import com.google.firebase.firestore.pipeline.FindNearestStage;
import com.google.firebase.firestore.pipeline.Ordering;
import com.google.firebase.firestore.pipeline.SampleStage;
import com.google.firebase.firestore.pipeline.Selectable;
import com.google.firebase.firestore.pipeline.UnnestOptions;
import java.util.List;
import java.util.Map;

/** Handles parsing and applying pipeline stages to Pipeline instances. */
class PipelineStageHandlers {
  private final ExpressionParsers parsers;

  PipelineStageHandlers(@NonNull ExpressionParsers parsers) {
    this.parsers = parsers;
  }

  /** Applies a pipeline stage to a Pipeline instance. */
  @SuppressWarnings("unchecked")
  Pipeline applyStage(
      @NonNull Pipeline pipeline,
      @NonNull String stageName,
      @Nullable Map<String, Object> args,
      @NonNull FirebaseFirestore firestore) {
    switch (stageName) {
      case "where":
        return handleWhere(pipeline, args);
      case "limit":
        return handleLimit(pipeline, args);
      case "offset":
        return handleOffset(pipeline, args);
      case "sort":
        return handleSort(pipeline, args);
      case "select":
        return handleSelect(pipeline, args);
      case "add_fields":
        return handleAddFields(pipeline, args);
      case "remove_fields":
        return handleRemoveFields(pipeline, args);
      case "distinct":
        return handleDistinct(pipeline, args);
      case "aggregate":
        return handleAggregate(pipeline, args);
      case "unnest":
        return handleUnnest(pipeline, args);
      case "replace_with":
        return handleReplaceWith(pipeline, args);
      case "union":
        return handleUnion(pipeline, args, firestore);
      case "sample":
        return handleSample(pipeline, args);
      case "find_nearest":
        return handleFindNearest(pipeline, args);
      default:
        throw new IllegalArgumentException("Unknown pipeline stage: " + stageName);
    }
  }

  private Pipeline handleWhere(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    Map<String, Object> expressionMap = (Map<String, Object>) args.get("expression");
    BooleanExpression booleanExpression = parsers.parseBooleanExpression(expressionMap);
    return pipeline.where(booleanExpression);
  }

  private Pipeline handleLimit(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    Number limit = (Number) args.get("limit");
    return pipeline.limit(limit.intValue());
  }

  private Pipeline handleOffset(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    Number offset = (Number) args.get("offset");
    return pipeline.offset(offset.intValue());
  }

  private Pipeline handleSort(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    Map<String, Object> orderingMap = (Map<String, Object>) args.get("expression");
    Expression expression = parsers.parseExpression(orderingMap);
    String direction = (String) args.get("order_direction");
    Ordering ordering = "asc".equals(direction) ? expression.ascending() : expression.descending();
    return pipeline.sort(ordering);
  }

  private Pipeline handleSelect(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    List<Map<String, Object>> expressionMaps = (List<Map<String, Object>>) args.get("expressions");

    if (expressionMaps == null || expressionMaps.isEmpty()) {
      throw new IllegalArgumentException("'select' requires at least one expression");
    }

    // Parse first expression as Selectable
    Selectable firstSelection = parsers.parseSelectable(expressionMaps.get(0));

    // Parse remaining expressions as varargs
    if (expressionMaps.size() == 1) {
      return pipeline.select(firstSelection);
    }

    Object[] additionalSelections = new Object[expressionMaps.size() - 1];
    for (int i = 1; i < expressionMaps.size(); i++) {
      Expression expr = parsers.parseExpression(expressionMaps.get(i));
      // Additional selections can be Selectable or any Object
      additionalSelections[i - 1] = expr;
    }

    return pipeline.select(firstSelection, additionalSelections);
  }

  private Pipeline handleAddFields(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    List<Map<String, Object>> expressionMaps = (List<Map<String, Object>>) args.get("expressions");

    if (expressionMaps == null || expressionMaps.isEmpty()) {
      throw new IllegalArgumentException("'add_fields' requires at least one expression");
    }

    // Parse first expression as Selectable
    Selectable firstField = parsers.parseSelectable(expressionMaps.get(0));

    // Parse remaining expressions as Selectable varargs
    if (expressionMaps.size() == 1) {
      return pipeline.addFields(firstField);
    }

    Selectable[] additionalFields = new Selectable[expressionMaps.size() - 1];
    for (int i = 1; i < expressionMaps.size(); i++) {
      additionalFields[i - 1] = parsers.parseSelectable(expressionMaps.get(i));
    }

    return pipeline.addFields(firstField, additionalFields);
  }

  private Pipeline handleRemoveFields(
      @NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    List<String> fieldPaths = (List<String>) args.get("field_paths");

    if (fieldPaths == null || fieldPaths.isEmpty()) {
      throw new IllegalArgumentException("'remove_fields' requires at least one field path");
    }

    // Convert first field path string to Field
    Field firstField = Expression.field(fieldPaths.get(0));

    // Convert remaining field paths to Field varargs
    if (fieldPaths.size() == 1) {
      return pipeline.removeFields(firstField);
    }

    Field[] additionalFields = new Field[fieldPaths.size() - 1];
    for (int i = 1; i < fieldPaths.size(); i++) {
      additionalFields[i - 1] = Expression.field(fieldPaths.get(i));
    }

    return pipeline.removeFields(firstField, additionalFields);
  }

  private Pipeline handleDistinct(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    List<Map<String, Object>> expressionMaps = (List<Map<String, Object>>) args.get("expressions");

    if (expressionMaps == null || expressionMaps.isEmpty()) {
      throw new IllegalArgumentException("'distinct' requires at least one expression");
    }

    // Parse first expression as Selectable
    Selectable firstGroup = parsers.parseSelectable(expressionMaps.get(0));

    // Parse remaining expressions as varargs (can be Selectable or Any)
    if (expressionMaps.size() == 1) {
      return pipeline.distinct(firstGroup);
    }

    Object[] additionalGroups = new Object[expressionMaps.size() - 1];
    for (int i = 1; i < expressionMaps.size(); i++) {
      Expression expr = parsers.parseExpression(expressionMaps.get(i));
      // Additional groups can be Selectable or any Object
      additionalGroups[i - 1] = expr;
    }

    return pipeline.distinct(firstGroup, additionalGroups);
  }

  @SuppressWarnings("unchecked")
  private Pipeline handleAggregate(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    // Check if this is using aggregate_stage (new API) or aggregate_functions (legacy API)
    if (args.containsKey("aggregate_stage")) {
      // New API: aggregateStage with optional options
      Map<String, Object> aggregateStageMap = (Map<String, Object>) args.get("aggregate_stage");
      AggregateStage aggregateStage = parsers.parseAggregateStage(aggregateStageMap);

      // Parse optional options
      Map<String, Object> optionsMap = (Map<String, Object>) args.get("options");
      if (optionsMap != null && !optionsMap.isEmpty()) {
        AggregateOptions options = parsers.parseAggregateOptions(optionsMap);
        return pipeline.aggregate(aggregateStage, options);
      } else {
        return pipeline.aggregate(aggregateStage);
      }
    } else {
      // Legacy API: aggregate_functions (varargs)
      List<Map<String, Object>> aggregateMaps =
          (List<Map<String, Object>>) args.get("aggregate_functions");

      if (aggregateMaps == null || aggregateMaps.isEmpty()) {
        throw new IllegalArgumentException(
            "'aggregate' requires at least one aggregate function or an aggregate_stage");
      }

      // Parse first aggregate function as AliasedAggregate
      AliasedAggregate firstAccumulator = parsers.parseAliasedAggregate(aggregateMaps.get(0));

      // Parse remaining aggregate functions as AliasedAggregate varargs
      if (aggregateMaps.size() == 1) {
        return pipeline.aggregate(firstAccumulator);
      }

      AliasedAggregate[] additionalAccumulators = new AliasedAggregate[aggregateMaps.size() - 1];
      for (int i = 1; i < aggregateMaps.size(); i++) {
        additionalAccumulators[i - 1] = parsers.parseAliasedAggregate(aggregateMaps.get(i));
      }

      return pipeline.aggregate(firstAccumulator, additionalAccumulators);
    }
  }

  private Pipeline handleUnnest(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    Map<String, Object> expressionMap = (Map<String, Object>) args.get("expression");
    Selectable expression = parsers.parseSelectable(expressionMap);
    String indexField = (String) args.get("index_field");
    if (indexField != null) {
      return pipeline.unnest(expression, new UnnestOptions().withIndexField(indexField));
    } else {
      return pipeline.unnest(expression);
    }
  }

  private Pipeline handleReplaceWith(
      @NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    Map<String, Object> expressionMap = (Map<String, Object>) args.get("expression");
    Expression expression = parsers.parseExpression(expressionMap);
    return pipeline.replaceWith(expression);
  }

  @SuppressWarnings("unchecked")
  private Pipeline handleUnion(
      @NonNull Pipeline pipeline,
      @Nullable Map<String, Object> args,
      @NonNull FirebaseFirestore firestore) {
    List<Map<String, Object>> nestedStages = (List<Map<String, Object>>) args.get("pipeline");
    if (nestedStages == null || nestedStages.isEmpty()) {
      throw new IllegalArgumentException("'union' requires a non-empty 'pipeline' argument");
    }
    Pipeline otherPipeline = PipelineParser.buildPipeline(firestore, nestedStages);
    return pipeline.union(otherPipeline);
  }

  private Pipeline handleSample(@NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    // Sample stage parsing
    Map<String, Object> sampleMap = (Map<String, Object>) args;
    // Parse sample configuration
    String type = (String) sampleMap.get("type");
    if (type == "percentage") {
      double value = (double) sampleMap.get("value");
      return pipeline.sample(SampleStage.withPercentage(value));
    } else {
      int value = (int) sampleMap.get("value");
      return pipeline.sample(SampleStage.withDocLimit(value));
    }
  }

  @SuppressWarnings("unchecked")
  private Pipeline handleFindNearest(
      @NonNull Pipeline pipeline, @Nullable Map<String, Object> args) {
    String vectorField = (String) args.get("vector_field");
    List<Number> vectorValue = (List<Number>) args.get("vector_value");
    String distanceMeasureStr = (String) args.get("distance_measure");
    Number limitObj = (Number) args.get("limit");

    if (distanceMeasureStr == null) {
      throw new IllegalArgumentException("'find_nearest' requires a 'distance_measure' argument");
    }

    // Convert Dart enum name to Android enum value
    FindNearestStage.DistanceMeasure distanceMeasure =
        parsers.parseDistanceMeasure(distanceMeasureStr);

    // Convert vector value to double array
    double[] vectorArray = new double[vectorValue.size()];
    for (int i = 0; i < vectorValue.size(); i++) {
      vectorArray[i] = vectorValue.get(i).doubleValue();
    }

    Field fieldExpr = Expression.field(vectorField);

    if (limitObj != null) {
      return pipeline.findNearest(
          vectorField,
          Expression.vector(vectorArray),
          distanceMeasure,
          new FindNearestOptions().withLimit(limitObj.intValue()));
    } else {
      return pipeline.findNearest(fieldExpr, vectorArray, distanceMeasure);
    }
  }
}

File: packages/cloud_firestore/cloud_firestore/android/src/main/kotlin/io/flutter/plugins/firebase/firestore/utils/PipelineStageHandlers.kt
/*
 * Copyright 2026, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.firestore.utils

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Pipeline
import com.google.firebase.firestore.pipeline.AggregateOptions
import com.google.firebase.firestore.pipeline.Expression
import com.google.firebase.firestore.pipeline.FindNearestOptions
import com.google.firebase.firestore.pipeline.Ordering
import com.google.firebase.firestore.pipeline.SampleStage
import com.google.firebase.firestore.pipeline.SearchStage
import com.google.firebase.firestore.pipeline.UnnestOptions

/** Handles parsing and applying pipeline stages to Pipeline instances. */
internal class PipelineStageHandlers(private val parsers: ExpressionParsers) {
  @Suppress("UNCHECKED_CAST")
  fun applyStage(
      pipeline: Pipeline,
      stageName: String,
      args: Map<String, Any>?,
      firestore: FirebaseFirestore
  ): Pipeline {
    return when (stageName) {
      "where" -> handleWhere(pipeline, args)
      "limit" -> handleLimit(pipeline, args)
      "offset" -> handleOffset(pipeline, args)
      "sort" -> handleSort(pipeline, args)
      "select" -> handleSelect(pipeline, args)
      "add_fields" -> handleAddFields(pipeline, args)
      "remove_fields" -> handleRemoveFields(pipeline, args)
      "distinct" -> handleDistinct(pipeline, args)
      "aggregate" -> handleAggregate(pipeline, args)
      "aggregate_with_options" -> handleAggregateWithOptions(pipeline, args)
      "unnest" -> handleUnnest(pipeline, args)
      "replace_with" -> handleReplaceWith(pipeline, args)
      "union" -> handleUnion(pipeline, args, firestore)
      "sample" -> handleSample(pipeline, args)
      "find_nearest" -> handleFindNearest(pipeline, args)
      "search" -> handleSearch(pipeline, args)
      else -> throw IllegalArgumentException("Unknown pipeline stage: $stageName")
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleWhere(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val expressionMap = args!!["expression"] as Map<String, Any>
    val booleanExpression = parsers.parseBooleanExpression(expressionMap)
    return pipeline.where(booleanExpression)
  }

  private fun handleLimit(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val limit = args!!["limit"] as Number
    return pipeline.limit(limit.toInt())
  }

  private fun handleOffset(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val offset = args!!["offset"] as Number
    return pipeline.offset(offset.toInt())
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleSort(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val orderingMaps = args!!["orderings"] as List<Map<String, Any>>?
    require(!orderingMaps.isNullOrEmpty()) { "'sort' requires at least one ordering" }

    val firstMap = orderingMaps[0]
    var expression = parsers.parseExpression(firstMap["expression"] as Map<String, Any>)
    var direction = firstMap["order_direction"] as String?
    val firstOrdering = if (direction == "asc") expression.ascending() else expression.descending()

    if (orderingMaps.size == 1) {
      return pipeline.sort(firstOrdering)
    }

    val additionalOrderings =
        Array(orderingMaps.size - 1) { i ->
          val map = orderingMaps[i + 1]
          expression = parsers.parseExpression(map["expression"] as Map<String, Any>)
          direction = map["order_direction"] as String?
          if (direction == "asc") expression.ascending() else expression.descending()
        }
    return pipeline.sort(firstOrdering, *additionalOrderings)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleSelect(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val expressionMaps = args!!["expressions"] as List<Map<String, Any>>?
    require(!expressionMaps.isNullOrEmpty()) { "'select' requires at least one expression" }

    val firstSelection = parsers.parseSelectable(expressionMaps[0])
    if (expressionMaps.size == 1) {
      return pipeline.select(firstSelection)
    }
    val additionalSelections =
        Array<Any>(expressionMaps.size - 1) { i -> parsers.parseExpression(expressionMaps[i + 1]) }
    return pipeline.select(firstSelection, *additionalSelections)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleAddFields(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val expressionMaps = args!!["expressions"] as List<Map<String, Any>>?
    require(!expressionMaps.isNullOrEmpty()) { "'add_fields' requires at least one expression" }

    val firstField = parsers.parseSelectable(expressionMaps[0])
    if (expressionMaps.size == 1) {
      return pipeline.addFields(firstField)
    }
    val additionalFields =
        Array(expressionMaps.size - 1) { i -> parsers.parseSelectable(expressionMaps[i + 1]) }
    return pipeline.addFields(firstField, *additionalFields)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleRemoveFields(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val fieldPaths = args!!["field_paths"] as List<String>?
    require(!fieldPaths.isNullOrEmpty()) { "'remove_fields' requires at least one field path" }

    val firstField = Expression.field(fieldPaths[0])
    if (fieldPaths.size == 1) {
      return pipeline.removeFields(firstField)
    }
    val additionalFields = Array(fieldPaths.size - 1) { i -> Expression.field(fieldPaths[i + 1]) }
    return pipeline.removeFields(firstField, *additionalFields)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleDistinct(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val expressionMaps = args!!["expressions"] as List<Map<String, Any>>?
    require(!expressionMaps.isNullOrEmpty()) { "'distinct' requires at least one expression" }

    val firstGroup = parsers.parseSelectable(expressionMaps[0])
    if (expressionMaps.size == 1) {
      return pipeline.distinct(firstGroup)
    }
    val additionalGroups =
        Array<Any>(expressionMaps.size - 1) { i -> parsers.parseExpression(expressionMaps[i + 1]) }
    return pipeline.distinct(firstGroup, *additionalGroups)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleAggregate(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val aggregateMaps = args!!["aggregate_functions"] as List<Map<String, Any>>?
    require(!aggregateMaps.isNullOrEmpty()) {
      "'aggregate' requires at least one aggregate function"
    }

    val firstAccumulator = parsers.parseAliasedAggregate(aggregateMaps[0])
    if (aggregateMaps.size == 1) {
      return pipeline.aggregate(firstAccumulator)
    }
    val additionalAccumulators =
        Array(aggregateMaps.size - 1) { i -> parsers.parseAliasedAggregate(aggregateMaps[i + 1]) }
    return pipeline.aggregate(firstAccumulator, *additionalAccumulators)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleAggregateWithOptions(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val aggregateStageMap = args!!["aggregate_stage"] as Map<String, Any>
    val aggregateStage = parsers.parseAggregateStage(aggregateStageMap)
    val optionsMap = args["options"] as Map<String, Any>?
    if (optionsMap != null && optionsMap.isNotEmpty()) {
      val options: AggregateOptions = parsers.parseAggregateOptions(optionsMap)
      return pipeline.aggregate(aggregateStage, options)
    }
    return pipeline.aggregate(aggregateStage)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleUnnest(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val expressionMap = args!!["expression"] as Map<String, Any>
    val expression = parsers.parseSelectable(expressionMap)
    val indexField = args["index_field"] as String?
    return if (indexField != null) {
      pipeline.unnest(expression, UnnestOptions().withIndexField(indexField))
    } else {
      pipeline.unnest(expression)
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleReplaceWith(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val expressionMap = args!!["expression"] as Map<String, Any>
    val expression = parsers.parseExpression(expressionMap)
    return pipeline.replaceWith(expression)
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleUnion(
      pipeline: Pipeline,
      args: Map<String, Any>?,
      firestore: FirebaseFirestore
  ): Pipeline {
    val nestedStages = args!!["pipeline"] as List<Map<String, Any>>?
    require(!nestedStages.isNullOrEmpty()) { "'union' requires a non-empty 'pipeline' argument" }
    val otherPipeline = PipelineParser.buildPipeline(firestore, nestedStages)
    return pipeline.union(otherPipeline)
  }

  private fun handleSample(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val sampleMap = args!!
    val type = sampleMap["type"] as String?
    return if (type == "percentage") {
      val value = (sampleMap["value"] as Number).toDouble()
      pipeline.sample(SampleStage.withPercentage(value))
    } else {
      val value = (sampleMap["value"] as Number).toInt()
      pipeline.sample(SampleStage.withDocLimit(value))
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleFindNearest(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    val vectorField = args!!["vector_field"] as String
    val vectorValue = args["vector_value"] as List<Number>
    val distanceMeasureStr = args["distance_measure"] as String?
    val limitObj = args["limit"] as Number?
    require(distanceMeasureStr != null) { "'find_nearest' requires a 'distance_measure' argument" }

    val distanceMeasure = parsers.parseDistanceMeasure(distanceMeasureStr)
    val vectorArray = DoubleArray(vectorValue.size) { i -> vectorValue[i].toDouble() }
    val fieldExpr = Expression.field(vectorField)
    return if (limitObj != null) {
      pipeline.findNearest(
          vectorField,
          Expression.vector(vectorArray),
          distanceMeasure,
          FindNearestOptions().withLimit(limitObj.toLong()))
    } else {
      pipeline.findNearest(fieldExpr, vectorArray, distanceMeasure)
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun handleSearch(pipeline: Pipeline, args: Map<String, Any>?): Pipeline {
    require(args != null) { "'search' requires arguments" }
    val queryType = args["query_type"] as String?
    val query = args["query"]
    var searchStage: SearchStage =
        when (queryType) {
          "string" -> SearchStage.withQuery(query as String)
          "expression" ->
              SearchStage.withQuery(parsers.parseBooleanExpression(query as Map<String, Any>))
          else ->
              throw IllegalArgumentException(
                  "'search' requires query_type to be either 'string' or 'expression'")
        }

    val sortMaps = args["sort"] as List<Map<String, Any>>?
    if (!sortMaps.isNullOrEmpty()) {
      val firstOrdering = parseOrdering(sortMaps[0])
      searchStage =
          if (sortMaps.size == 1) {
            searchStage.withSort(firstOrdering)
          } else {
            val additionalOrderings =
                Array(sortMaps.size - 1) { i -> parseOrdering(sortMaps[i + 1]) }
            searchStage.withSort(firstOrdering, *additionalOrderings)
          }
    }

    val addFieldMaps = args["add_fields"] as List<Map<String, Any>>?
    if (!addFieldMaps.isNullOrEmpty()) {
      val firstField = parsers.parseSelectable(addFieldMaps[0])
      searchStage =
          if (addFieldMaps.size == 1) {
            searchStage.withAddFields(firstField)
          } else {
            val additionalFields =
                Array(addFieldMaps.size - 1) { i -> parsers.parseSelectable(addFieldMaps[i + 1]) }
            searchStage.withAddFields(firstField, *additionalFields)
          }
    }

    val languageCode = args["language_code"] as String?
    if (languageCode != null) {
      searchStage = searchStage.withLanguageCode(languageCode)
    }
    val limit = args["limit"] as Number?
    if (limit != null) {
      searchStage = searchStage.withLimit(limit.toLong())
    }
    val offset = args["offset"] as Number?
    if (offset != null) {
      searchStage = searchStage.withOffset(offset.toLong())
    }
    val retrievalDepth = args["retrieval_depth"] as Number?
    if (retrievalDepth != null) {
      searchStage = searchStage.withRetrievalDepth(retrievalDepth.toLong())
    }
    return pipeline.search(searchStage)
  }

  @Suppress("UNCHECKED_CAST")
  private fun parseOrdering(orderingMap: Map<String, Any>): Ordering {
    val expression = parsers.parseExpression(orderingMap["expression"] as Map<String, Any>)
    val direction = orderingMap["order_direction"] as String?
    return if (direction == "asc") expression.ascending() else expression.descending()
  }
}
